### PR TITLE
Spring Boot CDS support

### PR DESCRIPTION
**Pull request by anthonydahanne**
* helper layer and run training run layer

published at: `anthonydahanne/spring-boot:app-cds`

### CODE REVIEW

File: helper/spring_cds.go [hunk 1]
```diff
@@ -32,25 +32,24 @@ type SpringCds struct {
 
 func (s SpringCds) Execute() (map[string]string, error) {
 
-	s.Logger.Info("Spring App CDS Enabled, contributing -XX:SharedArchiveFile=application.jsa to JAVA_OPTS")
 	s.Logger.Body("Those are the files we have in the workspace")
 
 	StartOSCommand("", "ls", "-al", "./")
 	StartOSCommand("", "env")
```
Comment (anthonydahanne) [inline helper/spring_cds.go:38]: could go away now

File: helper/spring_cloud_bindings.go [hunk 2]
```diff
@@ -34,30 +33,27 @@ type SpringCloudBindings struct {
...
 		enabled, err = strconv.ParseBool(val)
 		if err != nil {
 			return nil, fmt.Errorf("unable to parse $BPL_SPRING_CLOUD_BINDINGS_ENABLED\n%w", err)
 		}
 	}
-
 	// Switching from "BPL_SPRING_CLOUD_BINDINGS_ENABLED" to "BPL_SPRING_CLOUD_BINDINGS_DISABLED" which defaults to 'false' to follow convention
 	if sherpa.ResolveBool("BPL_SPRING_CLOUD_BINDINGS_DISABLED") || !enabled {
-		return nil, nil
+		opts = sherpa.AppendToEnvVar("JAVA_TOOL_OPTIONS", " ",  "-Dorg.springframework.cloud.bindings.boot.enable=false")
```
Comment (anthonydahanne) [inline helper/spring_cloud_bindings.go:50]: ah, the bug fix to disable SCB 👍

File: boot/build.go [hunk 3]
```diff
@@ -52,68 +57,50 @@ type Build struct {
...
 
-		// add labels
-		result.Labels, err = labels(context.Application.Path, manifest)
-		if err != nil {
-			return libcnb.BuildResult{}, err
+	version, version_found := manifest.Get("Spring-Boot-Version")
+	if !version_found {
+		if context.Application.Path, manifest, err = findSpringBootExecutableJAR(context.Application.Path); err != nil {
+			return libcnb.BuildResult{}, fmt.Errorf("unable to find Spring Boot Executable Jar\n%w", err)
+		} else if version, version_found = manifest.Get("Spring-Boot-Version"); !version_found {
```
Comment (anthonydahanne) [inline boot/build.go:82]: `} else {`

File: boot/build.go [hunk 4]
```diff
@@ -185,87 +155,120 @@ func (b Build) Build(context libcnb.BuildContext) (libcnb.BuildResult, error) {
...
-		// configure JVM for application type
-		classes, ok := manifest.Get("Spring-Boot-Classes")
-		if !ok {
-			return libcnb.BuildResult{}, fmt.Errorf("manifest does not contain Spring-Boot-Classes")
+		var scbVer string
+		var scbSet bool
+		if scbVer, scbSet = cr.Resolve("BP_SPRING_CLOUD_BINDINGS_VERSION"); !scbSet {
+			if scbVer, err = getSCBVersion(version); err != nil {
+				return libcnb.BuildResult{}, fmt.Errorf(
+					"unable to read the Spring Boot version from META-INF/MANIFEST.MF. " +
```
Comment (anthonydahanne) [inline boot/build.go:172]: this part of the error message is actually wrong. It should read:
"unable to parse the Spring Boot version in the META-INF/MANIFEST.MF. "

it's maybe a detail, but by this point, we were able to read the SB version for sure, we just did not recognize it

File: boot/build.go [hunk 5]
```diff
@@ -185,87 +155,120 @@ func (b Build) Build(context libcnb.BuildContext) (libcnb.BuildResult, error) {
...
+	}
+	at, err := NewWebApplicationType(context.Application.Path, wr)
+	if err != nil {
+		return libcnb.BuildResult{}, fmt.Errorf("unable to create WebApplicationType\n%w", err)
+	}
+	at.Logger = b.Logger
+	result.Layers = append(result.Layers, at)
+
+	// Slices
+	if index, ok := manifest.Get("Spring-Boot-Layers-Index"); ok {
```
Comment (anthonydahanne) [inline boot/build.go:259]: no slices for SpringCDS?

File: boot/spring_cds.go [hunk 6]
```diff
@@ -57,13 +60,33 @@ func NewSpringCds(cache libpak.DependencyCache, appPath string, manifest *proper
...
+			}
+		} else {
+			return layer, nil
+		}
+
+		// if unpack == true, follow this (with any updates needed), else (3.3+) run new command, +- touch file timestamps? + skip to training run section below
+
 		s.Logger.Bodyf("This is the value of AppPath: %s", s.AppPath)
 		s.Logger.Body("Those are the files we have in the workspace")
 		helper.StartOSCommand("", "ls", "-al", s.AppPath)
```
Comment (anthonydahanne) [inline boot/spring_cds.go:92]: can go away

**Review by dmikusa — COMMENTED** [review]
Questions.

File: boot/build.go [hunk 7]
```diff
@@ -369,12 +429,204 @@ func getSCBVersion (manifestVer string) (string, error) {
...
+	if err != nil {
+		return false
+	}
+	if bootThreeThreeConstraint.Check(bv) {
+		return true
+	}
+	return false
+}
+
+func Unzip(src, dest string) error {
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/build.go:573]: no, not really - I didn't know about it.

Now, we had a lot of issues with Jar zipping / unzipping (that I will explain in one of your other questions) and it was helpful to use our own Unzip to try few things out; I will try using crush and make sure all works fine.

File: boot/build.go [hunk 8]
```diff
@@ -369,12 +429,204 @@ func getSCBVersion (manifestVer string) (string, error) {
...
+			Type:      "web",
+			Command:   command,
+			Arguments: arguments,
+			Direct:    true,
+			Default:   true,
+		})
+	return processes
+}
+
+func (b *Build) findSpringBootExecutableJAR(appPath string) (string, *properties.Properties, error) {
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/build.go:510]: yeah, the naming of this function is not great.

When `spring-boot` BP `build.go`  is called, the content of `workspace` can be either a JAR (if maven / gradle built the jar) or source code (if you rely on `pack` to build it all).

In both cases, we need to know whether the jar is a Spring Boot jar or not; so once it's unzipped we can find out opening the `MANIFEST.MF`

Then,
* if we decide to do CDS: perfect, only one case to handle: exploded jar
* if we don't do CDS: cool, the exploded format is slightly faster


it should have been called `detectSpringBootExecutableJarAndExtract` - want me to rename it?

File: boot/spring_performance.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+			if err := CreateJar(s.AppPath+"/", tempJarPath); err != nil {
+				return layer, fmt.Errorf("error recreating jar\n%w", err)
+			}
+			f, err := os.Open(tempJarPath)
+			if err != nil {
+				return layer, fmt.Errorf("error opening jar\n%w", err)
+			}
+			if err = sherpa.CopyFile(f, filepath.Join(layer.Path, "runner.jar")); err != nil {
+				return layer, fmt.Errorf("error copying jar\n%w", err)
+			}
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/spring_performance.go:103]: that's because the Spring Boot feature that extracts into a CDS friendly format can not be called from an unzipped Spring Boot jar; see : https://github.com/spring-projects/spring-boot/issues/38276#issuecomment-1983451808

I know you may think : "what are they doing? unzipping, rezipping, and then creating a new layout? what a waste of IO!"

But: we have to make sure we deal with a spring boot jar (+1 for unzip) , apps run slightly faster when unzipped (+1 for unzip) preferably one single path when we get to `performance` (+1 for unzip), potentially add the SCB jar into the app (+1 for unzip) - and then we need to call spring boot tooling (+1 for rezip)

And, it happens at build time, so I don't think that playing with IO is a huge concern.

File: boot/spring_performance.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+		if err := s.Executor.Execute(effect.Execution{
+			Command: "java",
+			Env:     trainingRunEnvVariables,
+			Args:    trainingRunArgs,
+			Dir:     s.AppPath,
+			Stdout:  s.Logger.InfoWriter(),
+			Stderr:  s.Logger.InfoWriter(),
+		}); err != nil {
+			return libcnb.Layer{}, fmt.Errorf("error running build\n%w", err)
+		}
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/spring_performance.go:145]: well, we've run a whole lot of tests (at least 6 differentpeople with different JVM, hardware, OSes, etc.) and... that never was a concerned.

The "magic" is under this [Spring Framework 6 innovation](https://github.com/spring-projects/spring-framework/issues/30606): `-Dspring.context.exit=onRefresh` - when the beans are loaded, pouf, the app exits - and there's no way a working app would hung before that.

File: boot/spring_performance.go [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+	// we set the creation date to the buildpack default 1980/01/01 date; the date the layer will appear to be created at
+	return s.Executor.Execute(effect.Execution{
+		Command: "find",
+		Env:     []string{"TZ=UTC"},
+		Args:    []string{"./", "-exec", "touch", "-t", "198001010000.01", "{}", ";"},
+		Dir:     s.AppPath,
+		Stdout:  s.Logger.InfoWriter(),
+		Stderr:  s.Logger.InfoWriter(),
+	})
+}
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/spring_performance.go:166]: yeah, I started with the Go mtimes and unfortunately, I got slightly different results (CDS is super sensitive to timestamps) and since touch worked, and left if at that.

I can retry

File: go.mod [hunk 12]
```diff
@@ -7,28 +7,52 @@ require (
 	github.com/buildpacks/libcnb v1.30.3
 	github.com/heroku/color v0.0.6
 	github.com/magiconair/properties v1.8.7
+	github.com/mholt/archiver/v4 v4.0.0-alpha.8
 	github.com/onsi/gomega v1.33.0
 	github.com/paketo-buildpacks/libjvm v1.44.2
 	github.com/paketo-buildpacks/libpak v1.69.1
 	github.com/pelletier/go-toml v1.9.5
+	github.com/saracen/fastzip v0.1.11
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply go.mod]: no worries; they were just used during some testing / experimentation, they're now gone

File: internal/fsutil/walk.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+func Walk(root string, walkFn filepath.WalkFunc) error {
```
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply internal/fsutil/walk.go:11]: not sure... I guess I did not know about it?

trying out `fs.WalkDir(f, appPath, func(path string, info fs.DirEntry, err error)` now

File: boot/detect.go [hunk 14]
```diff
@@ -49,21 +53,40 @@ func (d Detect) Detect(context libcnb.DetectContext) (libcnb.DetectResult, error
...
+			Plans: []libcnb.BuildPlan{
+				{
+					Provides: []libcnb.BuildPlanProvide{
+						{Name: PlanEntrySpringBoot},
+					},
+					Requires: []libcnb.BuildPlanRequire{
+						{Name: PlanEntryJVMApplication},
+						{Name: PlanEntrySpringBoot},
+						// Require a JRE at build time to perform CDS training run
+						{Name: PlanEntryJRE, Metadata: map[string]interface{}{"build": true}},
```
[the remark this answers is not in the document]
Comment (dmikusa) [reply boot/detect.go:79]: Right, but you should have a JDK, at least if you're building from source code. I suspect if you're building from pre-compiled assets then you wouldn't have one, but I can't recall off the top of my head. That's why I was asking about that case specifically.

The reason I ask is because it seems like any time we make changes to the build plan, it's a whole thing. There's a very high degree of probability that it messes with something in unintended ways because the build plan is complicated, covering many different scenarios, and affecting all of the buildpacks not just this one (and not necessarily just Paketo buildpacks).

I don't know what the expectations or needs are here but I would offer a suggestion, that the buildpacks only attempt to generate the CDS training data if they are building from source code. 

If you were to do it this way. That would make things less complicated overall, it prevents you from needing to change the build plan, and I think that you could make a case that if someone is already building externally, which is more work for them, then they want more control and would probably also want more control over how the CDS training data is generated. In fact, if we don't support passing it in, I'm pretty confident that at some point we're going to get a request to allow this.

My suggestion in short, if you want the Easy Button then you build from source and build packs do it all. If you want to control how things are done, then you build and generate a CDS file and providing us the artifacts and we just make a container out of them.

Building with the Spring Boot build tools is something I don't know about though. That's definitely a use case to consider. I'm kind of assuming that there's a Maven/Gradle step that can do this, so when using that tool it could pass both the JAR and the CDS training file to buildpacks. If that's not the case, then it probably merits a discussion at least. Is that easy/hard to do? Do we think users might want to pre-generate this file and hand it off to buildpacks? etc...

File: internal/fsutil/walk.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply internal/fsutil/walk.go]: ok that worked! thanks!

File: boot/spring_performance.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (dmikusa) [reply boot/spring_performance.go:145]: I'm ok with it if you're confident in that it won't happen. A hung build can be a big deal for users though, for example it can be potentially costly since many people pay by the minute for CI/build time, so I just wanted to mention it. Make sure we're thinking through that scenario.

File: boot/spring_performance.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dmikusa) [reply boot/spring_performance.go:166]: I worry about this from a performance perspective. The shelling out to find isn't bad, it's one command. but find itself is going to shell out to touch for every file in that directory. This is the app workspace so we have an unbounded number of files that could potentially be in here. If someone has a lot of little files, maybe you end up with 10,000. That's a lot of commands to run.

Even if you can't get `os.Chtimes` to work, which it seems like it should be possible, you might look at using `find` and piping to `xargs`. That can be more efficient because `xargs` can batch multiple files into one call to touch. So if you have 10k files, maybe it can do 500 files per call to touch (depends on the max arg size which I don't remember off hand), but that cuts the number of calls to 1 find, 1 xargs, and 20 touch.

If you didn't want to do the piping, you could walk in Go, batch up the files and then do something similar to xargs calling touch with multiple files.

File: boot/spring_performance.go [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+		Stdout:  s.Logger.InfoWriter(),
+		Stderr:  s.Logger.InfoWriter(),
+	}); err != nil {
+		return fmt.Errorf("error extracting Jar with jarmode\n%w", err)
+	}
+	return nil
+}
+
+// CreateJar heavily inspired by: https://gosamples.dev/zip-file/
+func CreateJar(source, target string) error {
```
Comment (dmikusa) [inline boot/spring_performance.go:187]: If we have to keep this, I'd like to see it move into libpak under the crush module. It's going to keep things consistent with respect to handling archives. There's a CreateTar method in there, so seems reasonable to add a CreateZip or CreateJar method.

File: boot/detect.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/detect.go:79]: > that the buildpacks only attempt to generate the CDS training data if they are building from source code.
well, Spring Boot Gradle and Maven plugins do not do that; and I believe they're a pretty common way to consume the `spring-boot` buildpack; so I don't think we want to limit ourselves to source code only.

> then they want more control and would probably also want more control over how the CDS training data is generated

I don't agree with that: we expect most CDS users to be your average Spring Boot user, going to start.spring.io and then setting an env. variable (maybe it's even gonna be provided for them eventually) and obseve 1.5x faster startup times.
We don't believe users will want to tweak CDS; just don't use the Spring Boot buildpack if you're an expert?

> I'm kind of assuming that there's a Maven/Gradle step that can do this
Pretty sure there's not. I remember we asked them (@mhalbritter and @snicoll ) - not exactly that question, but on the topic of "please output something else than the spring boot jar or plain jar" and since it's an opinionated framework, they prefer limiting the possible outputs to the strict necessary.

File: boot/spring_performance.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (dmikusa) [reply boot/spring_performance.go:103]: I'm really just trying to understand everything, hence all the questions. I appreciate all the detailed responses.

I'd like to know more about what Boot is doing with this extraction command. In a nutshell, it sounds like it's extracting some stuff from the standard fat JAR and putting it into a different format (the comment made it seem like libs are pulled out and a "runner" jar is created). Since we already have the files extracted, it seems like we could move some files around on disk and arrive at the new format bypassing that command. Not knowing more about what Boot is doing, there may be details missing that might make this more complicated than it's worth.

I do think we need to consider the performance here. We're talking about the user's application, so it is an unbounded number of files. We could have 20M of files, we could have 2G of files so we can't just write this off as trivial. It could be a pretty noticeable performance impact for some users. The other point to consider is caching. I don't think you're going to be able to cache this very effectively. This is operating on the user's files, so you pretty much have to do this every time you build, since user's files could change every time. That means a performance hit is going to be felt on every build, not just first builds or occasionally when cache is invalidated.

Given that, I think we have to at least consider the level of effort for implementing this more efficiently. Maybe implementing the logic being done by the extract command in Go (or some other alternative solution), without all the extra file i/o and CPU. Maybe it's too complicated and we can't justify it but it seems worth at least discussing what that command is doing more or doing a spike to get a better handle on the decision.

-----------

Another concern of mine is the state of the user's application. 

1. When you build from their source, we get the user's source code in `/workspace`. 
    a. Maven/Gradle run and provide a Spring Boot JAR. The buildpack then extracts this, removes source code, and puts the compiled code into `/workspace`.
    b. Maven/Gradle run and provide multiple JAR files. The buildpack then takes all of these JAR files, removes source code, and puts all the JARs not extracted into `/workspace`.
2a. When the user provides a Spring Boot fat JAR, pack extracts this and puts it into `/workspace`.
2b. I think it fails at the moment with executable-jar, but a user can technically provide a directory of JAR files they pre-compiled, and pack will put set them as the contents of `/workspace`. This is probably a scenario that we'll support in the future though, there's been a desire to fix executable-jar in this regard.

It's the 1b.) and 2b.) scenarios that concern me. I think neither is probably super likely for a Boot app, but how are these situations going to be handled? Are we going to fail them as unsupported? Or are they legit and we need to handle them?

If we are not intending to handle them, I think that might simplify some of the logic around when you need to zip/unzip stuff.

File: boot/build.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/build.go]: ok, that seems to work fine with `crush` - replaced

File: boot/detect.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (dmikusa) [reply boot/detect.go:79]: >Pretty sure there's not. I remember we asked them (@mhalbritter and @snicoll ) - not exactly that question, but on the topic of "please output something else than the spring boot jar or plain jar" and since it's an opinionated framework, they prefer limiting the possible outputs to the strict necessary.

How would you do CDS without buildpacks then? Is the user left to sort out the Java commands? Maybe there isn't, but I've come to expect handy boot commands like `./gradlew bootRun` that just do things like this (and how native image builds just work). I was thinking something might exist for CDS too.

File: boot/spring_performance.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/spring_performance.go:103]: > I'm really just trying to understand everything, hence all the questions. I appreciate all the detailed responses.

👍

> what Boot is doing with this extraction command. In a nutshell, it sounds like it's extracting some stuff from the standard fat JAR and putting it into a different format

yeah something like that: extracts all jars in BOOT-INF/lib into their own folder, creates a jar with a "normal" java MANIFEST and all users classes and the list of jars in `BOOT-INF/lib`

>  Maybe implementing the logic being done by the extract command in Go

well, this is how we started with @pivotal-david-osullivan - but we reverted course because we don't want the `spring-boot` buildpack to "know" about internal Spring Boot transformation. I spend a lot of time actually trying to write a META-INF/MANIFEST and I can tell you, that's not just limited lines, weird indentation or whatnot; and there's no existing Go library that handles that.

It's a Spring Boot + Java transformation - I'm afraid Go is not the proper tooling to take care of that; a broken MANIFEST could hurt the user in weird ways...

[the remark this answers is not in the document]
Comment (anthonydahanne) [reply boot/spring_performance.go:103]: > It's the 1b.) and 2b.) scenarios that concern me. I think neither is probably super likely for a Boot app

right, there's no way a Spring Boot app would come as a bunch of jars; like I wrote a little bit earlier, Spring Boot team keeps the output options very limited

> If we are not intending to handle them, I think that might simplify some of the logic around when you need to zip/unzip stuff.

hum... do you mean we could remove the `findSpringBootExecutableJAR` function? at least the walking part? that does not seem like a huge gain.

File: boot/detect.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (mhalbritter) [reply boot/detect.go:79]: That's correct, there's no support on the Maven or Gradle plugins for this. You can [read more in the documentation](https://docs.spring.io/spring-boot/reference/deployment/efficient.html#deployment.efficient.cds).

[the remark this answers is not in the document]
Comment (pivotal-david-osullivan) [reply boot/detect.go:79]: While I am definitely familiar with the woes of touching the build-plan as Dan mentions, I made this change (when I realised we had no `java` available to run the training-run command), somewhat safe (?) in the knowledge that:

* it is behind the check for opting-in to CDS via the variable - which doesn't remove the risk of issues but narrows the scope
* tl;dr - I believe overall we are expanding plan requirement rather than mixing across current states, the latter being more prone to problems I think. 
  * As far as I could tell nothing else currently requires a JRE at build-time - tomcat/liberty etc require it at run time for obvious reasons. This results in this change **adding** to the Metadata for the JRE-at-runtime entry that buildpacks such as Executable Jar request by default: `Buildpack Plan: {Entries:[{Name:jre Metadata:map[launch:true]} {Name:jre Metadata:map[build:true]}]}`. If building from source, the same appears alongside the build-system buildpack's requested JDK `Name:jdk Metadata:map[]`

[the remark this answers is not in the document]
Comment (dmikusa) [reply boot/detect.go:79]: Sounds good @pivotal-david-osullivan. I agree that this seems lower risk than some of the changes in the past, if for no other reason than that users have to opt-in before they would be impacted.

> As far as I could tell nothing else currently requires a JRE at build-time - tomcat/liberty etc require it at run time for obvious reasons.

+1 I can't think of anything that does either.

File: boot/spring_performance.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pivotal-david-osullivan) [reply boot/spring_performance.go]: I added a commit that swaps this call out for `Chtimes` and it works!

File: boot/spring_performance.go (hunk 15 above)
Comment (anthonydahanne) [reply boot/spring_performance.go]: done, using libpak 1.7.0
